File: packages/js-sdk/src/runCode.ts
import { Sandbox } from './sandbox'

export enum CodeRuntime {
  Node16 = 'Node16',
  Python3 = 'Python3',
  Bash = 'Bash',
  Python3_DataAnalysis = 'Python3-DataAnalysis',
}

/**
 * Run code in a sandboxed cloud playground.
 * `runCode` wraps the `Sandbox` class and provides a simple interface for running code in a sandboxed environment
 * without any need to manage lifecycle of the sandbox.
 * `runCode` automatically loads the E2B API key from the `E2B_API_KEY` environment variable.
 *
 * @param runtime The runtime to use when running the code. Can be one of the following:
 * - "Node16"
 * - "Python3"
 *
 * **Let us know if you need support for other runtimes.**
 * @param code The code to run
 * @param opts Optional parameters to pass
 * @returns
 */
export async function runCode(
  runtime: CodeRuntime,
  code: string,
  opts?: { apiKey?: string },
) {
  let binary = ''
  let filepath = ''
  let envID = ''
  switch (runtime) {
    case CodeRuntime.Node16:
      envID = 'base'
      binary = 'node'
      filepath = '/index.js'
      break
    case CodeRuntime.Python3:
      envID = 'base'
      binary = 'python3'
      filepath = '/main.py'
      break
    case CodeRuntime.Python3_DataAnalysis:
      envID = 'Python3-DataAnalysis'
      binary = 'python3'
      filepath = '/main.py'
      break
    case CodeRuntime.Bash:
      envID = 'base'
      binary = 'bash'
      filepath = '/main.sh'
      break
    default:
      throw new Error(
        `The "${runtime}" runtime isn't supported. Please contact us (hello@e2b.dev) if you need support for this runtime`,
      )
  }

  const sandbox = await Sandbox.create({
    template: envID,
    apiKey: opts?.apiKey || process?.env?.E2B_API_KEY || '', // Sandbox.create will throw an error if the API key is not provided so no need to check here
  })

  await sandbox.filesystem.write(filepath, code)

  const out = await sandbox.process.startAndWait(`${binary} ${filepath}`)

  await sandbox.close()

  return {
    stdout: out.stdout,
    stderr: out.stderr,
  }
}

File: packages/python-sdk/e2b/sandbox/run_code.py
from os import getenv
from typing import Literal, Union, Optional

from . import Sandbox
from .exception import UnsupportedRuntimeException

CodeRuntime = Literal[
    "Node16",
    "Python3",
    "Bash",
    "Python3-DataAnalysis",
]


def run_code(
    runtime: Union[CodeRuntime, str],
    code: str,
    api_key: Optional[str] = None,
):
    """
    Run code in a sandboxed cloud playground and return the stdout and stderr.

    `run_code` wraps the `Sandbox` class and provides a simple interface for running code in a sandbox
    without any need to manage lifecycle of the sandbox.
    `run_code` automatically loads the E2B API key from the `E2B_API_KEY` environment variable.

    :param runtime: The runtime to run the code in. One of "Node16" or "Python3".
    :param code: The code to run
    :param api_key: The E2B API key to use. If not provided, the `E2B_API_KEY` environment variable is used.

    :return: A string touple of stdout and stderr
    """

    if api_key is None:
        api_key = getenv("E2B_API_KEY")

    if runtime == "Node16":
        env_id = "base"
        binary = "node"
        filepath = "/index.js"
    elif runtime == "Python3":
        env_id = "base"
        binary = "python3"
        filepath = "/index.py"
    elif runtime == "Python3-DataAnalysis":
        env_id = "Python3-DataAnalysis"
        binary = "python3"
        filepath = "/index.py"
    elif runtime == "Bash":
        env_id = "base"
        binary = "bash"
        filepath = "/main.sh"
    else:
        raise UnsupportedRuntimeException(
            f'Invalid runtime "{runtime}". Please contact us (hello@e2b.dev) if you need support for this runtime'
        )

    sandbox = Sandbox(id=env_id, api_key=api_key)
    sandbox.filesystem.write(filepath, code)

    proc = sandbox.process.start(cmd=f"{binary} {filepath}")
    proc.wait()

    sandbox.close()

    return proc.output.stdout, proc.output.stderr
